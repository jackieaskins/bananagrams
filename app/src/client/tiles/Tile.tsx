import { Box, Text } from "@chakra-ui/react";
import { useDrag } from "react-dnd";
import { Tile as TileType } from "@/types/tile";

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

export default function Tile({
  boardLocation,
  tile: { id, letter },
  color = "black",
}: TileProps): JSX.Element {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: "TILE", id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = boardLocation ? "0" : "5px";

  return (
    <Box
      ref={dragRef}
      alignItems="center"
      backgroundColor="yellow.100"
      border="1px"
      borderColor={color}
      borderRadius="5px"
      color={color}
      cursor="move"
      display="inline-flex"
      height="25px"
      justifyContent="center"
      margin={margin}
      opacity={isDragging ? 0.5 : 1}
      width="25px"
    >
      <Text as="b">{letter}</Text>
    </Box>
  );
}
