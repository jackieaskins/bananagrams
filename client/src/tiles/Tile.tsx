/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Typography, Box } from '@material-ui/core';
import { useDrag } from 'react-dnd';

import { Tile as TileType } from './types';

type TileProps = {
  boardLocation: { x: number; y: number } | null;
  tile: TileType;
  color?: string;
};

const Tile = ({
  boardLocation,
  tile: { id, letter },
  color = 'black',
}: TileProps): JSX.Element => {
  const [{ isDragging }, dragRef] = useDrag({
    item: { type: 'TILE', id, boardLocation },
    collect: (monitor) => ({
      isDragging: monitor.isDragging(),
    }),
  });

  const margin = !!boardLocation ? '0' : '5px';

  return (
    <Box
      // @ts-ignore
      ref={dragRef}
      display="inline-flex"
      justifyContent="center"
      alignItems="center"
      border={1}
      borderRadius="borderRadius"
      style={{
        color: color,
        backgroundColor: '#ffffc7',
        cursor: 'move',
        height: '23px',
        margin,
        opacity: isDragging ? 0.5 : 1,
        width: '23px',
      }}
    >
      <Typography variant="body2">{letter}</Typography>
    </Box>
  );
};

export default Tile;
