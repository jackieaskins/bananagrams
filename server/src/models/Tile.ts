import BaseModel from './BaseModel';

export type TileJSON = {
  id: string;
  letter: string;
};

export default class Tile implements BaseModel<TileJSON> {
  private id: string;
  private letter: string;

  constructor(id: string, letter: string) {
    this.id = id;
    this.letter = letter;
  }

  getId(): string {
    return this.id;
  }

  getLetter(): string {
    return this.letter;
  }

  toJSON(): TileJSON {
    const { id, letter } = this;
    return { id, letter };
  }

  reset(): void {
    return;
  }
}
