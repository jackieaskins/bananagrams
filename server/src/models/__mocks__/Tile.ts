export default jest.fn().mockImplementation((id, letter) => ({
  getId: jest.fn().mockReturnValue(id),
  getLetter: jest.fn().mockReturnValue(letter),
  toJSON: jest.fn().mockReturnValue(`TileJSON - ${id} ${letter}`),
}));
