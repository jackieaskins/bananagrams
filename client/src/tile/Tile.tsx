import { memo, useCallback } from 'react';
import { useRecoilValue } from 'recoil';

import { Tile as TileType } from '../tiles/types';
import { TILE_SIZE } from './constants';
import { selectedTileState } from './selectedTileState';

interface TileProps extends TileType {
  onClick?: ({ tile }: { tile: TileType }) => void;
}

const Tile = ({ onClick, id, letter }: TileProps): JSX.Element => {
  const selectedTile = useRecoilValue(selectedTileState);
  const isSelectedTile = selectedTile?.tile?.id === id;

  const handleClick = useCallback(() => {
    onClick?.({ tile: { id, letter } });
  }, [id, letter, onClick]);

  return (
    <div
      className="no-drag tile"
      css={{
        width: `${TILE_SIZE}px`,
        height: `${TILE_SIZE}px`,
        backgroundColor: '#ffffc7',
        border: '0.5px solid black',
        borderRadius: '5px',
        flexShrink: 0,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        opacity: isSelectedTile ? 0.5 : 1,
      }}
      onClick={handleClick}
    >
      {letter}
    </div>
  );
};
export default memo(Tile);
