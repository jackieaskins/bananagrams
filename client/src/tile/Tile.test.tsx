import { shallow } from 'enzyme';
import { useRecoilValue } from 'recoil';

import Tile from './Tile';

const mockUseRecoilValue = useRecoilValue as jest.Mock;
jest.mock('recoil', () => ({
  useRecoilValue: jest.fn(),
}));

jest.mock('./selectedTileState', () => ({
  selectedTileState: 'mockSelectedTileState',
}));

describe('<Tile />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<Tile letter="A" id="A1" {...propOverrides} />);

  it('renders styled div', () => {
    expect(renderComponent()).toMatchSnapshot();
  });

  it('renders at half opacity if tile is selected', () => {
    mockUseRecoilValue.mockReturnValue({ tile: { id: 'A1', letter: 'A' } });

    expect(renderComponent()).toMatchSnapshot();
  });

  it('calls onClick on click', () => {
    const mockOnClick = jest.fn();

    renderComponent({ onClick: mockOnClick }).simulate('click');

    expect(mockOnClick).toHaveBeenCalledWith({
      tile: { id: 'A1', letter: 'A' },
    });
  });

  it('handles onClick not being passed', () => {
    expect(() => renderComponent().simulate('click')).not.toThrow();
  });
});
