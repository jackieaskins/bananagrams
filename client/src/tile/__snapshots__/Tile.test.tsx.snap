// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tile /> renders at half opacity if tile is selected 1`] = `
.emotion-0ezjqdi-Tile {
  width: 30px;
  height: 30px;
  background-color: #ffffc7;
  border: 0.5px solid black;
  border-radius: 5px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  opacity: 1;
}

.emotion-012n29wf-Tile {
  width: 30px;
  height: 30px;
  background-color: #ffffc7;
  border: 0.5px solid black;
  border-radius: 5px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  opacity: 0.5;
}

<div
  className="no-drag tile emotion-0"
  onClick={[Function]}
>
  A
</div>
`;

exports[`<Tile /> renders styled div 1`] = `
.emotion-0ezjqdi-Tile {
  width: 30px;
  height: 30px;
  background-color: #ffffc7;
  border: 0.5px solid black;
  border-radius: 5px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  opacity: 1;
}

<div
  className="no-drag tile emotion-0"
  onClick={[Function]}
>
  A
</div>
`;
